kolibri/plugins/coach/assets/src/state/actions/exam.js

```diff
@@ -271,25 +271,67 @@ function showExamReportDetailPage(store, classId, examId) {
   store.dispatch('CORE_SET_PAGE_LOADING', false);
 }
 
-function activateExam() {
-  this.displayModal(false);
+function activateExam(store, examId) {
+  ExamResource.getModel(examId).save({ active: true }).then(
+    () => {
+      const exams = store.state.pageState.exams;
+      const examIndex = exams.findIndex(exam => exam.id === examId);
+      exams[examIndex].active = true;
+
+      store.dispatch('SET_EXAMS', exams);
```

Comment:
One pattern that we have in the app is a sequence of mutations within an action, e.g. set page loading, set page name, set page state, set page title.

One thing we've noticed occasionally is that the UI is liable to be updated after each mutation. This has caused some difficulties when the state tree is momentarily in an inconsistent state - e.g. the page name has updated, but the page state still refers to the old page. This is partly why we have "set loading false" as sort of mutex around all the other mutations - it hides the vue completely and prevents things from trying to render an inconsistent state.

When we migrate to vuex 2, I was definitely interested in looking more closely at what logic belongs inside the mutation, and what belongs in the action. (Note the names have changed - I'm using the vuex 1 names.) In particular, I'd like to write our mutation such that we can avoid these intermediary view updates and get rid of the 'page loading' mutex hack.